### ISSUE TITLE
THNN unnecessary .zero() calls on gradients

### ISSUE THREAD

**Issue by szagoruyko** [pytorch/pytorch#393]
THNN convolutions are slower due to unnecessary .zero() calls on gradients:
* https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/conv.py#L227 is not needed
* https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/conv.py#L239 same for bias, prob worth adding a constant in C-side that controls whether to override grad or not

**Comment by szagoruyko** [comment]
We should reopen this, still many functions have unnecessary .zero() calls on grad_input and grad_param, e.g.:
* `conv2d`: https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/conv.py#L245
* `batch_norm`: https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/batchnorm.py#L53
* most of the `auto` functions don't need it https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/thnn/auto.py#L163

**Comment by colesbury** [comment]
I think the `grad_weight` ones are currently necessary.

**Comment by szagoruyko** [comment]
@colesbury for nn yes, but cudnn has:
* `alpha` and `beta` on cudnnConvolutionBackwardFilter
* `alphaParamDiff` and `betaParamDiff` on cudnnBatchNormalizationBackward
which control addition of gradients and should be set to 1 and 0 in pytorch cudnn (it is `scale` and 1 in cudnn.torch: https://github.com/soumith/cudnn.torch/blob/master/SpatialConvolution.lua#L265)

**Comment by colesbury** [comment]
Yeah, we can remove it on BatchNorm. If you look at Conv, it already only zeros out grad_weight for cunn.

**Comment by soumith** [comment]
this is now fixed in master: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/batch_normalization.cpp#L119
